pkg/kubelet/kubelet_pods.go

```diff
@@ -1831,6 +1854,71 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 			oldStatusPtr = &oldStatus
 		}
 		status := convertContainerStatus(cStatus, oldStatusPtr)
+		if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
```

Comment:
This makes sense. There's a [pre-beta TODO](https://github.com/kubernetes/enhancements/issues/1287#issue-503823628) to rework this part. Lantao has a good point that we should store the CRI ContainerStatus.Resources instead of converting to v1 resources in toKubeContainerStatus function. I'll add another TODO here to makes sure it's not missed. This is behaving more like generate code so it can be pulled into a generate function and that fn can be called. Doesn't look too hard.

Reply:
To be clear, this is blocking feedback. Adding this here breaks the non-mutating guarantees of this function, I don't think this can be taken as a TODO.

/hold

Reply:
> The additional nuance is to populate the resourcesAllocated, which comes from the checkpoint store which is not technically a conversion, but the code groups together well with the data being populated.

This change cannot happen here. Updates need to happen later. This should only be populating a data structure and calculating existing state, not making changes to state. This is a core architectural issue with this implementation.

Reply:
> > The additional nuance is to populate the resourcesAllocated, which comes from the checkpoint store which is not technically a conversion, but the code groups together well with the data being populated.
> 
> This change cannot happen here. Updates need to happen later. This should only be populating a data structure and calculating existing state, not making changes to state. This is a core architectural issue with this implementation.

And what **not making changes to state** in this block of code? Which specific line(s) of code?

Reply:
I am not seeing any changes to state in present form of this function.

A follow-up for me is ensuring that status.Resources() actually includes all resources, not just cpu,memory,ephemeral-storage, but also includes any extended resources that may have been present on the pod spec.

**TODO for Derek** see if there is a test case here that includes extended resources.

Reply:
I have not handled displaying extended resources in the status. Does it need to be part of this KEP? If yes, can it be handled in as a beta blocker?

Reply:
ensuring extended resources are properly handled can be a beta blocker.

Reply:
Tracking issue https://github.com/kubernetes/kubernetes/issues/114159